### PR TITLE
Transform script

### PR DESCRIPTION
**Pull request by lcawood**
Added a `transform.py` script which is to form part of the `pipeline.py` script. In addition, I have added a `test_transform.py` script which includes a suite of tests which aim to handle data that is missing or containing errors. 

The test suite is relatively comprehensive and all tests contained in `test_transform.py` pass, although it's important to note that we are likely to discover new ways for the data to contain errors or be incomplete once the pipeline is up and running. We can address these issues, if (and most likely when) they arise later. 

The `transform.py` script has a pylint score of 10.

### CODE REVIEW

File: pipeline/pipeline.py [hunk 1]
```diff
@@ -94,4 +94,5 @@ def pipeline():
 
 
 if __name__ == "__main__":
-    pipeline()
+    # pipeline()
```
Comment (Shurc1val) [inline pipeline/pipeline.py:97]: Why is this commented out?

File: pipeline/transform.py [hunk 2]
```diff
@@ -1,32 +1,223 @@
-"""Module to transform the data received from the Kafka cluster"""
+'''Module to transform the data received from the Kafka cluster.'''
```
Comment (Shurc1val) [inline pipeline/transform.py:1]: ''' to """ for consistency across files, I think?

File: pipeline/transform.py [hunk 3]
```diff
@@ -1,32 +1,223 @@
...
+
+
+def timestamp_to_date(timestamp_ms: int) -> str:
+    '''Helper function that converts a timestamp in milliseconds
+    since the Unix epoch to a date string in the form YYYY-MM-DD.'''
+
+    return datetime.utcfromtimestamp(timestamp_ms / 1000).strftime('%Y-%m-%d')
+
+
+def check_datetime_is_valid(dt: datetime) -> bool:
```
Comment (Shurc1val) [inline pipeline/transform.py:19]: Nice touch

File: pipeline/database_functions.py [hunk 4]
```diff
@@ -1,4 +1,35 @@
...
+            password=environ['db_password'],
+            host=environ['db_host'],
+            port=environ['db_port']
+        )
+    except OperationalError as e:
+        print(f"The error '{e}' occurred.")
+    finally:
+        return conn
+
+
```
Comment (CharliedeanCS) [inline pipeline/database_functions.py:27]: This all needs to be removed. As the database connection is setup already.

File: pipeline/transform.py [hunk 5]
```diff
@@ -1,32 +1,223 @@
...
+        except ValueError:
+            return None
+        return None
+    return None
+
+
+def get_bike_serial_number_from_log_line(log_line: str) -> str | None:
+    '''Takes in a kafka log line, and returns the bike serial number if found.'''
+
+    log_line_dict = literal_eval(log_line.split('=')[1])
```
Comment (Shurc1val) [inline pipeline/transform.py:49]: Tidy

File: pipeline/test_transform.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
+# pylint: skip-file
```
Comment (CharliedeanCS) [inline pipeline/test_transform.py:1]: Why are you skipping this file?

File: pipeline/test_transform.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+from datetime import datetime
+
+from transform import timestamp_to_date, check_datetime_is_valid, extract_datetime_from_string, \
+    get_bike_serial_number_from_log_line, get_email_from_log_line, get_user_from_log_line, \
+    get_ride_data_from_log_line, get_reading_data_from_log_line, get_address_from_log_line
+
+
+def test_convert_epoch_to_datetime():
```
Comment (CharliedeanCS) [inline pipeline/test_transform.py:9]: All functions should have doc strings to explain what the function is doing.

File: pipeline/transform.py [hunk 8]
```diff
@@ -1,32 +1,223 @@
...
+
+
+def get_email_from_log_line(log_line: str) -> str | None:
+    '''Helper function to extract an email address from a log line using regex if found.'''
+
+    email_pattern = r'[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}'
+    match = re.search(email_pattern, log_line)
+    if match:
+        return match.group()
+    return None
```
Comment (Shurc1val) [inline pipeline/transform.py:62]: would be easier literal eval again rather than a regex

File: terraform/main.tf
Comment (Shurc1val) [inline terraform/main.tf:1]: Good job Louie!

File: pipeline/transform.py [hunk 9]
```diff
@@ -1,32 +1,223 @@
...
+    if match:
+        try:
+            datetime_obj = datetime.strptime(
+                match.group(), '%Y-%m-%d %H:%M:%S.%f')
+            if check_datetime_is_valid(datetime_obj):
+                return datetime_obj
+        except ValueError:
+            return None
+        return None
+    return None
```
[the remark this answers is not in the document]
Comment (Shurc1val) [reply pipeline/transform.py]: think we're going to have some issues with regex for postcodes and nlp for addresses in transform.py, as a lot of the details look ficticious; however, they do appear consistent in structure, following "{address line 1}, {address line 2}, {city}, {postcode}" , where address line 2 may or may not be present, so this shouldn't be too hard a fix.

**Review by Shurc1val — APPROVED** [review]
Looks good to me!

**Review by Shurc1val — APPROVED** [review]
Continues to look fine; we shall see on integration with the whole pipeline...

**Pull request merged**